R/zchunk_LA154.transportation_UCD.R

```diff
@@ -40,116 +44,376 @@ module_energy_LA154.transportation_UCD_DISABLED <- function(command, ...) {
+      group_by(iso, UCD_sector, mode, year) %>%
+      summarise(value = sum(value))
+
+    # Spreading by variable to join all at once
+    UCD_trn_data_variable_spread <- UCD_trn_data_fillout %>%
+      ungroup %>%
+      distinct() %>%
+      select(-unit) %>%
+      spread(variable, value)
+
+    ALL_ctry_var <- IEA_histfut_data_times_UCD_shares %>%
+      left_join_error_no_match(UCD_ctry, by = "iso") %>%
+      # The energy weights will be replaced by the energy weights of each mode, as many techs have 0 consumption in the base year
+      select(-value) %>%
+      fast_left_join(IEA_data_times_UCD_shares_UCD_sector_agg,
+                     by = c("iso", "UCD_sector", "mode", "year")) %>%
+      # Using a floor on the weighting factor to avoid having zero weights for any countries
+      mutate(weight_EJ = pmax(value, energy.MIN_WEIGHT_EJ)) %>%
+      # Next, match in the derived variables, specific to each individual country/sector/mode/size.class/tech/fuel, except speed
+      # There will be NA non-fuel costs which can be set to zero
+      fast_left_join(UCD_trn_data_variable_spread %>% select(-speed),
+                     by = c("UCD_sector", "mode", "size.class", "UCD_technology", "UCD_fuel", "year", "UCD_region")) %>%
+      replace_na(list(`non-fuel costs` = 0))
+
+    # Adding in speed - this is matched by the mode and (for some) size class. Match size class first
+    speed_data <- UCD_trn_data_variable_spread %>%
+      select(UCD_sector, mode, size.class, year, UCD_region, speed) %>%
+      filter(!is.na(speed))
+
+    ALL_ctry_var <- ALL_ctry_var %>%
+      fast_left_join(speed_data,
+                     by = c("UCD_sector", "mode", "size.class", "year", "UCD_region")) %>%
+      # For the missing values, join using the mode ID
+      left_join_keep_first_only(speed_data,
+                                by = c("UCD_sector", "mode", "year", "UCD_region")) %>%
+      mutate(speed.x = if_else(is.na(speed.x), speed.y, speed.x)) %>%
+      replace_na(list(speed.x = 1))
+
+    # Calculate the weighted volumes. No need to convert units because these will be aggregated and divided by weights
+    ALL_region_var <- ALL_ctry_var %>%
+      mutate(Tvkm = weight_EJ / intensity) %>%
+      mutate(Tpkm = Tvkm * `load factor`) %>%
+      mutate(Tusd = Tvkm * `non-fuel costs`) %>%
+      mutate(Thr = Tvkm / speed.x ) %>%
+      group_by(UCD_technology,UCD_fuel, UCD_sector, mode, size.class.x, year, GCAM_region_ID) %>%
+      summarise(weight_EJ = sum(weight_EJ), Tvkm = sum(Tvkm), Tpkm = sum(Tpkm),Tusd = sum(Tusd), Thr = sum(Thr))
+
+    # Reverse the calculations to calculate the weighted average of each derived variable
+    ALL_region_var <- ALL_region_var %>%
+      mutate(intensity_MJvkm = weight_EJ / Tvkm,
+             loadfactor = Tpkm / Tvkm,
+             cost_usdvkm = Tusd / Tvkm,
+             speed_kmhr = Tvkm / Thr) %>%
+      gather(variable, value, 8:16) %>%
```

Comment:
Again, we'd rather not depend on specific column numbers. Is this gathering years? Then please use the `matches(YEAR_PATTERN)` construct above.